### PR TITLE
DAOS-4257 iosrv: move vos enumerate to object module

### PR DESCRIPTION
**Pull request by wangdi1**
Move vos enumerate to object module, so object client
module can use it for EC object enumeration.

And Some cleanup and fixes for enumeration process.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Comment by wangdi1** [comment]
The failure is due to 5325

File: src/object/obj_enum.c [hunk 1]
```diff
@@ -387,7 +386,8 @@ fill_rec(daos_handle_t ih, vos_iter_entry_t *key_ent, struct dss_enum_arg *arg,
 
 		d_iov_set(&iov_out, iovs[arg->sgl_idx].iov_buf +
 				       iovs[arg->sgl_idx].iov_len, data_size);
-		rc = vos_iter_copy(ih, key_ent, &iov_out);
+		D_ASSERT(arg->copy_data_cb != NULL);
+		rc = arg->copy_data_cb(ih, key_ent, &iov_out);
```
Comment (NiuYawei) [inline src/object/obj_enum.c:390]: Why don't call VOS API directly?

File: src/object/obj_enum.c [hunk 2]
```diff
@@ -468,18 +469,17 @@ enum_pack_cb(daos_handle_t ih, vos_iter_entry_t *entry, vos_iter_type_t type,
...
 {
 	int	rc;
 
 	D_ASSERT(!arg->fill_recxs ||
 		 type == VOS_ITER_SINGLE || type == VOS_ITER_RECX);
 
-	rc = vos_iterate(param, type, recursive, anchors, enum_pack_cb, NULL,
-			 arg, dth);
+	rc = iter_cb(param, type, recursive, anchors, enum_pack_cb, NULL,
+		     arg, dth);
```
Comment (NiuYawei) [inline src/object/obj_enum.c:480]: Why don't call VOS API directly?

File: src/rdb/rdb_raft.c [hunk 3]
```diff
@@ -404,12 +404,13 @@ rdb_raft_pack_chunk(daos_handle_t lc, struct rdb_raft_is *is, d_iov_t *kds,
...
 	arg.sgl = &sgl;
 
+	arg.copy_data_cb = vos_iter_copy;
 	/* Attempt to inline all values until recx bulks are implemented. */
 	arg.inline_thres = 1 * 1024 * 1024;
 
 	/* Enumerate from the object level. */
 	rc = dss_enum_pack(&param, VOS_ITER_OBJ, true, &anchors, &arg,
-			   NULL /* dth */);
+			   vos_iterate, NULL /* dth */);
```
Comment (NiuYawei) [inline src/rdb/rdb_raft.c:413]: I think we don't want RDB to call object API, probably that's why we put the dss_enum_pack() in iosrv at the very beginning?

File: src/object/obj_enum.c (hunk 1 above)
Comment (wangdi1) [reply src/object/obj_enum.c:390]: object client will use this file. but object client can not use VOS api.

File: src/object/obj_enum.c (hunk 2 above)
Comment (wangdi1) [reply src/object/obj_enum.c:480]: see above

File: src/rdb/rdb_raft.c (hunk 3 above)
Comment (wangdi1) [reply src/rdb/rdb_raft.c:413]: not sure the reason behind it, but we need call these object streaming api for client object.

File: src/include/daos_srv/daos_server.h [hunk 4]
```diff
@@ -722,10 +725,20 @@ struct dss_enum_arg {
...
+
 int dss_enum_pack(vos_iter_param_t *param, vos_iter_type_t type, bool recursive,
 		  struct vos_iter_anchors *anchors, struct dss_enum_arg *arg,
-		  struct dtx_handle *dth);
-
+		  enum_iterate_cb_t iter_cb, struct dtx_handle *dth);
+typedef int (*obj_enum_process_cb_t)(daos_key_desc_t *kds, void *ptr,
+				     unsigned int size, void *arg);
+int
+obj_enum_iterate(daos_key_desc_t *kdss, d_sg_list_t *sgl, int nr,
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/include/daos_srv/daos_server.h:739]: yes, I will do that in the following patches.

**Comment by gnailzenh** [comment]
I saw we got 5 failures, are they all known issues?

**Comment by wangdi1** [comment]
Yes, all of them are related with recent mercury changes and DAOS-5311

File: src/object/obj_enum.c [hunk 5]
```diff
@@ -468,18 +469,17 @@ enum_pack_cb(daos_handle_t ih, vos_iter_entry_t *entry, vos_iter_type_t type,
 int
 dss_enum_pack(vos_iter_param_t *param, vos_iter_type_t type, bool recursive,
 	      struct vos_iter_anchors *anchors, struct dss_enum_arg *arg,
```
Comment (johannlombardi) [inline src/object/obj_enum.c:471]: We should not have functions prefixed with dss_ in src/object. Please renamed those.

**Pull request merged**